### ISSUE TITLE
ReOpen: NetTcp and Kerberos in Linux

### ISSUE THREAD

**Issue by Cronan** [dotnet/wcf#2335]
I've re-opened due to this failing when called from unmanaged code in Linux.
@mconnew 

When I call the following code:

```
var binding = new NetTcpBinding();
var endpoint = new EndpointAddress(new Uri("net.tcp://someuri:8184/myInterface"));
var factory = new ChannelFactory<MyInterface>(binding, endpoint);
factory.Credentials.Windows.ClientCredential = new System.Net.NetworkCredential(@"my_username", "my_password");
var channel = factory.CreateChannel();
var result = channel.MyFunction(1);
```

I get the following error in Linux:

> Unhandled Exception: System.ServiceModel.EndpointNotFoundException: No IPEndpoints were found for host someuri.
>    at System.ServiceModel.Channels.SocketConnectionInitiator.<ConnectAsync>d__9.MoveNext()

Is NetTcp supported in Linux?

**Comment by Cronan** [comment]
Managed to get this working, amazing stuff, even calls out to Kerberos

```
var binding = new NetTcpBinding();
var endpoint = new EndpointAddress(new Uri("net.tcp://someuri:8184/myInterface"), 
    EndpointIdentity.CreateSpnIdentity("tcp/someuri:8184/myInterface"));
var factory = new ChannelFactory<MyInterface>(binding, endpoint);
factory.Credentials.Windows.ClientCredential = 
    new System.Net.NetworkCredential(@"my_username", "my_password");
var channel = factory.CreateChannel();
var result = channel.MyFunction(1);
```

**Comment by mcblair** [comment]
Have you considered implementing this in .NET Core 2.0 inside an AWS Lambda function?

It's not functioning due to `System.Reflections.DispatchProxy` being missing.

Locally it's not functioning due to `System.Private.ServiceModel` being missing. I've been scrubbing the web for two days and there is barely a peep on this stuff.

**Comment by Cronan** [comment]
@mcblair After allowing `dotnet msbuild` to restore packages, I found `System.Private.ServiceModel` here:
`~/my_app/build/lib.linux-x86_64-2.7/netcoreapp2.0/runtimes/unix/lib/netstandard2.0`

**Comment by Cronan** [comment]
I've re-opened due to this failing when called from unmanaged code in Linux.
To confirm, this works when run using `dotnet run` from a console app, but fails when called from unmanaged code.

```
var binding = new NetTcpBinding();
var endpoint = new EndpointAddress(new Uri("net.tcp://someuri:8184/myInterface"), 
    EndpointIdentity.CreateSpnIdentity("tcp/someuri:8184"));
var factory = new ChannelFactory<MyInterface>(binding, endpoint);
factory.Credentials.Windows.ClientCredential = 
    new System.Net.NetworkCredential(@"my_username", "my_password");
var channel = factory.CreateChannel();
var result = channel.MyFunction(1);
```
The exceptions I'm getting are as follows:
```
System.ServiceModel.Security.SecurityNegotiationException: A call to SSPI failed, see inner exception.
   at System.ServiceModel.Channels.WindowsStreamSecurityUpgradeProvider.WindowsStreamSecurityUpgradeInitiator.<OnInitiateUpgradeAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.StreamSecurityUpgradeInitiatorBase.<InitiateUpgradeAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.ConnectionUpgradeHelper.<InitiateUpgradeAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.ClientFramingDuplexSessionChannel.<SendPreambleAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.ConnectionPoolHelper.<EstablishConnectionAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.ClientFramingDuplexSessionChannel.<OnOpenAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.CommunicationObject.<OnOpenAsyncInternal>d__80.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.CommunicationObject.<System-ServiceModel-IAsyncCommunicationObject-OpenAsync>d__79.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.CommunicationObject.<OpenAsyncInternal>d__78.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.CommunicationObject.Open(TimeSpan timeout)
   at System.Runtime.TaskHelpers.<CallActionAsync>d__17`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.CommunicationObject.<OpenOtherAsync>d__109.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.ServiceChannel.<OnOpenAsync>d__151.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.CommunicationObject.<OnOpenAsyncInternal>d__80.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.CommunicationObject.<System-ServiceModel-IAsyncCommunicationObject-OpenAsync>d__79.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.CommunicationObject.<OpenAsyncInternal>d__78.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.ServiceModel.Channels.ServiceChannel.CallOpenOnce.System.ServiceModel.Channels.ServiceChannel.ICallOnce.Call(ServiceChannel channel, TimeSpan timeout)
   at System.ServiceModel.Channels.ServiceChannel.CallOnceManager.CallOnce(TimeSpan timeout, CallOnceManager cascade)
   at System.ServiceModel.Channels.ServiceChannel.EnsureOpened(TimeSpan timeout)
   at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)
   at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(MethodCall methodCall, ProxyOperationRuntime operation)
   at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(MethodInfo targetMethod, Object[] args)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Reflection.DispatchProxyGenerator.Invoke(Object[] args)
   at generatedProxy_1...(client code methods removed)
```
The inner exception mentioned above is as follows:
```
{Interop+NetSecurityNative+GssApiException: GSSAPI operation failed with error - The referenced credential has expired (Unknown code 0).
   at System.Net.Security.NegotiateStreamPal.GssInitSecurityContext(SafeGssContextHandle& context, 
SafeGssCredHandle credential, Boolean isNtlm, SafeGssNameHandle targetName, GssFlags inFlags, Byte[] buffer, Byte[]& outputBuffer, UInt32& outFlags, Int32& isNtlmUsed)}

HResult: 720896
Minor Status: 100005
Source: "System.Net.Security"
```

**Comment by Cronan** [comment]
@iamjasonp Sorry to be a pain, but I think you did a lot of work on the [](url)

**Comment by Cronan** [comment]
Closed and logged as #2585
